giskard/llm/evaluators/string_matcher.py

```diff
@@ -0,0 +1,103 @@
+import re
+import string
+import logging
+from typing import Tuple, Dict, List
+from dataclasses import dataclass
+
+from .base import BaseEvaluator, EvaluationResult
+from ...datasets.base import Dataset
+from ...models.base.model import BaseModel
+from ..errors import LLMGenerationError
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+
+@dataclass(frozen=True)
+class StringMatchingMethod:
+    substrings: Tuple[str]
+    all_substrings_must_be_found: bool = True
+    exact_matching: bool = False
+    word_matching: bool = False
```

Comment:
Now that I fully read the code, usage of flags seems inappropriate in this case because they are exclusive. I propose to instead set it as a string:
`method="exact"`, `method="word"`, `method="substring"` (you can probably find better names.

Just a simple string. Then in the code you would have:
```py3
if method == "exact":
    ...
elif method == "word":
    ...
else:
    raise ValueError(f"Invalid method '{method}'. Accepted values are exact, word, ...")
```

In my opinion, a simple string would be more explicit and clean.

The others (`case_sensitive`, `punctuation_sensitive`) are clear, because they are real flags that you can enable independently.

Reply:
> because they are exclusive.

They are not **exclusive**

it's true that if `exact_matching==True` then `word_matching` becomes obsolete, but if `exact_matching==False`, you can either opt for `word_matching==True` or not. Thus the need for flags.
```py
    if evaluator_config.exact_matching:
        return substring == prediction
    if evaluator_config.word_matching:
        return re.search(r"\b" + re.escape(substring) + r"\b", prediction) is not None
    return substring in prediction
```